### PR TITLE
Catalog: Ensure epoch does not decrease

### PR DESCRIPTION
**Pull request by jkosh44**
Previously migrating from one catalog implementation to another
allowed the catalog epoch to decrease. The problem was that we were
treating the epoch as an internal implementation detail. During the
migration from the stash to persist, we would NOT copy the epoch from
the stash to persist. Instead, persist would start with a brand-new
epoch at 2 (the starting value is 2 for historical reasons). The
problem is that the epoch is exposed externally, and it's used by the
storage controller. The storage controller uses the epoch for its own
separate fencing purposes. If the epoch decreases after a migration,
then the storage controller would mistakenly think it was being fenced
out by a higher epoch and fail to boot.

This commit adds an epoch lower bound to catalog opening. When the
catalog opens, it ensures that it's epoch is at least as big as the
epoch lower bound. During a migration, we pass the old epoch plus one
as the epoch lower bound, which ensure that the epoch never decreases.

Works towards resolving #24218

### Motivation
This PR fixes a previously unreported bug.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/catalog/src/durable/impls/persist.rs [hunk 1]
```diff
@@ -199,6 +201,9 @@ impl UnopenedPersistCatalogState {
         // Only writable catalogs attempt to increment the epoch.
         if matches!(mode, Mode::Writable) {
             current_epoch = current_epoch + 1;
+            if let Some(epoch_lower_bound) = epoch_lower_bound {
```
Comment (danhhz) [inline src/catalog/src/durable/impls/persist.rs:204]: might be worth putting an info log here? should be pretty high signal low noise. ditto in the other impl

File: src/stash/src/postgres.rs [hunk 2]
```diff
@@ -628,12 +643,13 @@ impl Stash {
...
                 // connection that also bumps it once).
-                let row = tx
-                    .query_one(
-                        "UPDATE fence SET epoch=epoch+1, nonce=$1 RETURNING epoch",
-                        &[&self.nonce.to_vec()],
-                    )
-                    .await?;
+                let update = match epoch_lower_bound {
+                    Some(epoch_lower_bound) => {
+                        format!("UPDATE fence SET epoch=GREATEST(epoch+1, {}), nonce=$1 RETURNING epoch", epoch_lower_bound.get())
```
Comment (danhhz) [inline src/stash/src/postgres.rs:648]: very optional: could map None to the minimum Epoch to only have one query here

**Review by danhhz — APPROVED** [review]
LGTM but my review was a bit scan-y, so you might want someone from adapter to take a look too

File: src/stash/src/postgres.rs (hunk 2 above)
Comment (mjibson) [inline src/stash/src/postgres.rs:648]: (I have no opinion on the single query idea.)

But I do think this should not use string formatting and instead change params to a vec that can have varying sizes if needed.

**Review by mjibson — APPROVED** [review]

File: src/catalog/src/durable/impls/persist.rs (hunk 1 above)
Comment (jkosh44) [reply src/catalog/src/durable/impls/persist.rs]: Done

File: src/stash/src/postgres.rs (hunk 2 above)
Comment (jkosh44) [reply src/stash/src/postgres.rs]: Done

Comment (jkosh44) [reply src/stash/src/postgres.rs]: > change params to a vec that can have varying sizes if needed.

I don't think I fully understand what this means. Does https://github.com/MaterializeInc/materialize/pull/24290/commits/bd133024b18f55fe59dcf5642dc206f548406113 accomplish this?

File: src/stash/src/postgres.rs [hunk 3]
```diff
@@ -628,12 +647,8 @@ impl Stash {
...
                 // current epoch has been bumped exactly once, then gets recreated by another
                 // connection that also bumps it once).
-                let row = tx
-                    .query_one(
-                        "UPDATE fence SET epoch=epoch+1, nonce=$1 RETURNING epoch",
-                        &[&self.nonce.to_vec()],
-                    )
-                    .await?;
+                let epoch_lower_bound = epoch_lower_bound.unwrap_or(NonZeroI64::MIN).get();
+                let row = tx.query_one(&format!("UPDATE fence SET epoch=GREATEST(epoch+1, $1), nonce=$2 RETURNING epoch"), &[&epoch_lower_bound, &self.nonce.to_vec()]).await?;
```
Comment (mjibson) [inline src/stash/src/postgres.rs:651]:
```suggestion
                let row = tx.query_one("UPDATE fence SET epoch=GREATEST(epoch+1, $1), nonce=$2 RETURNING epoch", &[&epoch_lower_bound, &self.nonce.to_vec()]).await?;
```

**Review by mjibson — APPROVED** [review]

File: src/stash/src/postgres.rs (hunk 2 above)
Comment (mjibson) [reply src/stash/src/postgres.rs]: Yes.

File: src/stash/src/postgres.rs (hunk 3 above)
Comment (jkosh44) [reply src/stash/src/postgres.rs:651]: Weird that the linter missed this, it usually catches these.

**Pull request merged**